geoutils/raster/delayed.py

```diff
@@ -0,0 +1,746 @@
+        list_subsamples = [
+            _delayed_subsample_block(b, ind)
+            for i, (b, ind) in enumerate(zip(blocks, ind_per_block))
+            if len(ind_per_block[i]) > 0
+        ]
+        # Cast output to the right expected dtype and length, then compute and concatenate
+        list_subsamples_delayed = [
+            da.from_delayed(s, shape=(nb_valids_per_block[i]), dtype=darr.dtype) for i, s in enumerate(list_subsamples)
+        ]
+        subsamples = np.concatenate(dask.compute(*list_subsamples_delayed), axis=0)
+
+        return subsamples
+
+    # To return indices
+    else:
+        # Get robust list of starts (using what is done in block_id of dask.array.map_blocks)
+        # https://github.com/dask/dask/blob/24493f58660cb933855ba7629848881a6e2458c1/dask/array/core.py#L908
+        from dask.utils import cached_cumsum
+
+        starts = [cached_cumsum(c, initial_zero=True) for c in darr.chunks]
+        num_chunks = darr.numblocks
+        # Get the starts per 1D block ID by unravelling starting indexes for each block
+        indexes_xi, indexes_yi = np.unravel_index(np.arange(len(blocks)), shape=(num_chunks[0], num_chunks[1]))
+        block_ids = [
+            {"xstart": starts[0][indexes_xi[i]], "ystart": starts[1][indexes_yi[i]]} for i in range(len(blocks))
+        ]
+
+        # Task delayed subsample indices to be computed for each block, skipping blocks with no values to sample
+        list_subsample_indices = [
+            _delayed_subsample_indices_block(b, ind, block_id=block_ids[i])
+            for i, (b, ind) in enumerate(zip(blocks, ind_per_block))
+            if len(ind_per_block[i]) > 0
+        ]
+        # Cast output to the right expected dtype and length, then compute and concatenate
+        list_subsamples_indices_delayed = [
+            da.from_delayed(s, shape=(2, len(ind_per_block[i])), dtype=np.dtype("int32"))
+            for i, s in enumerate(list_subsample_indices)
+        ]
+        indices = np.concatenate(dask.compute(*list_subsamples_indices_delayed), axis=0)
+
+        return indices[:, 0], indices[:, 1]
+
+
+# 2/ POINT INTERPOLATION ON REGULAR OR EQUAL GRID
+# This functionality is not covered efficiently by Dask/Xarray, because they need to support rectilinear grids, which
+# is difficult when interpolating in the chunked dimensions, and loads nearly all array memory when using .interp().
+
+# Here we harness the fact that rasters are always on regular (or sometimes equal) grids to efficiently map
+# the location of the blocks required for interpolation, which requires little memory usage.
+
+# Code structure inspired by https://blog.dask.org/2021/07/02/ragged-output and the "block_id" in map_blocks
+
+
+def _get_interp_indices_per_block(interp_x, interp_y, starts, num_chunks, chunksize, xres, yres):
+    """Map blocks where each pair of interpolation coordinates will have to be computed."""
+
+    # TODO 1: Check the robustness for chunksize different and X and Y
```

Comment:
I was testing this function out and unfortunately ran into a `IndexError`. 

 ```shell
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[3], [line 8](vscode-notebook-cell:?execution_count=3&line=8)
      [4](vscode-notebook-cell:?execution_count=3&line=4) xs, ys = np.unravel_index(samples, shape=dask_array.shape)
      [6](vscode-notebook-cell:?execution_count=3&line=6) print(xs,ys)
----> [8](vscode-notebook-cell:?execution_count=3&line=8) output = delayed_interp_points(darr=dask_array, points=(xs, ys), resolution=(0.3,0.3))

File [~/projects/xdem/geoutils/geoutils/raster/delayed.py:336](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:336), in delayed_interp_points(darr, points, resolution, method)
    [333](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:333) num_chunks = expanded.numblocks
    [335](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:335) # Get samples indices per blocks
--> [336](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:336) ind_per_block = _get_interp_indices_per_block(
    [337](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:337)     points[0, :], points[1, :], starts, num_chunks, chunksize, resolution[0], resolution[1]
    [338](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:338) )
    [340](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:340) # Create a delayed object for each block, and flatten the blocks into a 1d shape
    [341](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:341) blocks = expanded.to_delayed().ravel()

File [~/projects/xdem/geoutils/geoutils/raster/delayed.py:265](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:265), in _get_interp_indices_per_block(interp_x, interp_y, starts, num_chunks, chunksize, xres, yres)
    [260](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:260) for i, (x, y) in enumerate(zip(interp_x, interp_y)):
    [261](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:261)     # Because it is a regular grid, we know exactly in which block ID the coordinate will fall
    [262](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:262)     block_i_1d = int((x - starts[0][0]) [/](https://file+.vscode-resource.vscode-cdn.net/) (xres * chunksize[0])) * num_chunks[1] + int(
    [263](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:263)         (y - starts[1][0]) [/](https://file+.vscode-resource.vscode-cdn.net/) (yres * chunksize[1])
    [264](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:264)     )
--> [265](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:265)     ind_per_block[block_i_1d].append(i)
    [267](https://file+.vscode-resource.vscode-cdn.net/home/amelie/projects/xdem/geoutils/examples/~/projects/xdem/geoutils/geoutils/raster/delayed.py:267) return ind_per_block

IndexError: list index out of range
``` 

I first thought it was because of the uneven chunksizes of my array. As far as I understood you are assuming a regular grid here. I think it would be beneficial to make this robust for chunks that are of varying sizes (at least around the edges).  I'm assuming it's a common use case that a raster doesn't perfectly fit the chunksize.  

However! The actual error also appeared when I was testing with a small 4x4 array and the chunksize was 2 (and the chunks are even) ... not sure I completely understand how `block_i_1d` is calculated so I can't say for sure why it's running into the error but to help reproduce, this is my input data: 

```python
data = np.array(
        [
            [1, 2, 3, 4],
            [5, 6, 7, 8],
            [9, 10, 11, 12],
            [13, 14, 15, 16],
        ]
    )
dask_array = da.from_array(data, chunks=(chunk, chunk))

samples = np.random.default_rng().choice(
    dask_array.shape[0] * dask_array.shape[1],
    size=10,
    replace=False,
)
xs, ys = np.unravel_index(samples, shape=dask_array.shape)


output = delayed_interp_points(
    darr=dask_array,
    points=np.array([xs, ys]).reshape((-1, 2)),
    resolution=(0.3, 0.3),
)
```
 
I think using `DaskArray.chunksize` is a bit problematic here because it returns the max of `self.chunks` per dimension. For example: 

```python
dask_array = da.from_array(data, chunks=((1,1,2),(2,2)))
dask_array.chunksize
>>> (2,2)
```

Reply:
I think your error might be coming from the fact that `delayed_interp_points` expects coordinates (not indexes like `delayed_subsample`), and your specified resolution is 0.3, so your 4x4 array coordinates are 0 to 1.2 and some of your coordinates have a value of 2 or 3.

To generate random coordinates, you can use the same routine as what is currently done in `test_delayed_interp_points` (which also tests for uneven chunksizes and passes successfully :wink:)! Feel free to add any parameters/inputs to the tests if you noticed I missed something, I explain the logic in the delayed tests at the start of `TestDelayed` and in the description of each `test_delayed_xxx`.

When this function will be wrapped in the Xarray accessor `interp_points` function (like `interp_points` is in the `Raster` class right now), there will be a list of input checks, one of them verifying which input points belongs in the array and returning NaN for those who don't with a warning (and running only the ones within the array extent)! Ideally these checks should not live in the subfunction (depends if we make `delayed_interp_points` public, or just the wrapper function in the Xarray accessor), which is why I didn't add them here :smile:

Reply:
For the Xarray accessor, it will be finalizing this: #446 (the issue in Xarray was just to ask for approval/keep them posted, no code needs to live there), but we'll likely open a small PR simply to add our accessor in their doc to direct interested users here :wink: .